dht/sharder.hh

```diff
@@ -28,28 +28,76 @@
+// one (or a few) at a time, rather than accumulating them.
+
+// A mapping between a partition_range and a shard. All positions within `ring_range` are
+// owned by `shard`.
+//
+// The classes that return ring_position_range_and_shard make `ring_range` as large as
+// possible (maximizing the number of tokens), so the total number of such ranges is minimized.
+// Successive ranges therefore always have a different `shard` than the previous return.
+// (classes that return ring_position_range_and_shard_and_element can have the same `shard`
+// in successive returns, if `element` is different).
 struct ring_position_range_and_shard {
     dht::partition_range ring_range;
     unsigned shard;
 };
 
+// Incrementally divides a `partition_range` into sub-ranges wholly owned by a single shard.
 class ring_position_range_sharder {
     const sharder& _sharder;
     dht::partition_range _range;
     bool _done = false;
 public:
+    // Initializes the ring_position_range_sharder with a given range to subdivide.
     ring_position_range_sharder(const sharder& sharder, nonwrapping_range<ring_position> rrp)
             : _sharder(sharder), _range(std::move(rrp)) {}
+    // Fetches the next range-shard mapping. When the input range is exhausted, std::nullopt is
+    // returned. The returned ranges are contiguous and non-overlapping, and together span the
+    // entire input range.
     std::optional<ring_position_range_and_shard> next(const schema& s);
 };
 
+// A mapping between a partition_range and a shard (like ring_position_range_and_shard) extended
+// by having a reference to input range index. See ring_position_range_vector_sharder for use.
+//
+// The classes that return ring_position_range_and_shard_and_element make `ring_range` as large as
+// possible (maximizing the number of tokens), so the total number of such ranges is minimized.
+// Successive ranges therefore always have a different `shard` than the previous return.
+// (classes that return ring_position_range_and_shard_and_element can have the same `shard`
+// in successive returns, if `element` is different).
 struct ring_position_range_and_shard_and_element : ring_position_range_and_shard {
     ring_position_range_and_shard_and_element(ring_position_range_and_shard&& rpras, unsigned element)
             : ring_position_range_and_shard(std::move(rpras)), element(element) {
     }
     unsigned element;
 };
 
+// Incrementally divides several non-overlapping `partition_range`:s into sub-ranges wholly owned by
+// a single shard.
+//
+// Similar to ring_position_range_sharder, but instead of stopping when the input range is exhauseted,
+// moves on to the next input range (input ranges are supplied in a vector).
+//
+// This has two use cases:
+
+// 1. vnodes. A vnode cannot be described by a single range, since
+//    one vnode wraps around from the largest token back to the smallest token. Hence it must be
```

Comment:
I would say "may wrap" instead of "wraps". Obviously not all token ranges wrap around.